ONSdigital/rm-actionexporter-service
# Fix address used in action request

## Do not merge yet: Need to ensure all related PRs are tested in Pre-prod and discussed with FWM Team first

# Motivation and Context
We are not using the most recent address when a print file is being generated for social reminders. We therefore need to ensure the most recent address is used in action requests, and also to allow business action requests to still process correctly as a result of these changes.

Deploying this change will require the app to be stopped immediately after starting up and running the liquibase changes for the data migration script to be run manually as it may take too long to complete during app start up.

# What has changed
* Tweaked logic to ensure address is always persisted, if present.
* Added liquibase schema changes and migration script to be run manually (we will need to test the migration handling)
* Updated models.
* Added integration test to ensure the most recent address is used in the action request.

Related PRs:
[rm-actionsvc-api](https://github.com/ONSdigital/rm-actionsvc-api/pull/34)
[rm-action-service](https://github.com/ONSdigital/rm-action-service/pull/116)

# How to test?
* Checkout above branches from the related PRs.
* Ensure unit and integration tests pass.
* Acceptance tests should still pass (use `make setup` and `make rasrm_acceptance_tests`)

To test the migration:
* Run `make pull up` in docker dev to have all services running on latest master
* Get some data in the action exporter service (by executing collection exercises with associated print type action plans or acceptance tests)
* Re build the docker images on these branches
* Run `make up` once the branches are built, actionsvc and actionexportersvc containers should be recreated off the branch builds
* Tail the actionexportersvc logs. Once it successfully starts, stop or pause the service
* Run the data migration script in this PR (stored with the liquibase release scripts) then restart the service
* Check all the action requests in the actionexportersvc now have an addressFK corresponding to the correct addressPK row (for existing business addresses these rows may be all nulls, this is normal)
* Run a collection exercise with print type action plans and check the action exporter service generates the correct print files (particularly for social type which should include address details)

The travis build is expected to fail until the actionsvc-api changes are merged.

# Links
[Trello](https://trello.com/c/BNWyG8P1/260-bug200-addresses-are-not-being-updated-when-a-new-one-for-that-ru-is-being-updated-20)
https://github.com/ONSdigital/rm-action-service/pull/116
https://github.com/ONSdigital/rm-actionsvc-api/pull/34
